libshortfin/src/shortfin/process/worker.cc

```diff
@@ -0,0 +1,158 @@
+// Copyright 2024 Advanced Micro Devices, Inc
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "shortfin/process/worker.h"
+
+#include "shortfin/support/logging.h"
+
+namespace shortfin {
+
+Worker::Worker(const Options options)
+    : options_(std::move(options)),
+      signal_transact_(false),
+      signal_ended_(false) {
+  // Set up loop.
+  auto OnError = +[](void* self, iree_status_t status) {
+    // TODO: FIX ME.
+    iree_status_fprint(stderr, status);
+    iree_status_ignore(status);
+  };
+  iree_loop_sync_options_t loop_options = {.max_queue_depth = 128,
+                                           .max_wait_count = 64};
+  SHORTFIN_THROW_IF_ERROR(
+      iree_loop_sync_allocate(loop_options, options_.allocator, &loop_sync_));
+  iree_loop_sync_scope_initialize(loop_sync_, OnError, this, &loop_scope_);
+  loop_ = iree_loop_sync_scope(&loop_scope_);
+}
+
+Worker::~Worker() {
+  iree_loop_sync_scope_deinitialize(&loop_scope_);
+  iree_loop_sync_free(loop_sync_);
+  thread_.reset();
+}
+
+iree_status_t Worker::TransactLoop(iree_status_t signal_status) {
+  if (!iree_status_is_ok(signal_status)) {
+    // TODO: Handle failure.
+    return signal_status;
+  }
+
+  logging::info("Transact loop!");
+
+  {
+    // An outside thread cannot change the state we are managing without
+    // entering this critical section, so it is safe to reset the event
+    // here (it is not possible for it to be spurious reset).
+    iree_slim_mutex_lock_guard guard(mu_);
+    signal_transact_.reset();
+    if (kill_) {
+      // TODO: Do we want to somehow hard abort loop in flight work (vs
+      // just stopping submission of new work)?
+      return iree_ok_status();
+    }
+    next_thunks_.swap(pending_thunks_);
+  }
+
+  // Handle all callbacks.
+  for (auto& next_thunk : next_thunks_) {
+    // TODO: Make thunks have to return a status, propagate, and handle
+    // exceptions.
+    next_thunk();
```

Comment:
would be worth noting on EnqueueCallback that these aren't re-entrant - or change the loop to handle that (no iterator that'll cause explosions if next_thunks_t is modified while iterating)

Reply:
What am I missing: I think these are re-entrant. EnqueueCallbacks pushes onto pending_thunks_. Then when the loop next transacts, it swaps pending_thunks_ -> next_thunks_ (at the instance level to avoid conserve allocations). At least that was the intent. I don't think next_thunks_ can be modified during iteration.